lib/app_profiler/middleware/upload_action.rb

```diff
@@ -16,8 +21,43 @@ def call(profile, response: nil, autoredirect: nil)
           )
         end
 
+        def enqueue_upload(profile)
+          @queue ||= init_queue
+          begin
+            @queue.push(profile, true) # non-blocking push, raises ThreadError if queue is full
+          rescue ThreadError
+            AppProfiler.logger.info("[AppProfiler] upload queue is full, profile discarded")
+          end
+        end
+
+        def start_process_queue_thread
+          @process_queue_thread ||= Thread.new do
+            loop do
+              process_queue
+              sleep(AppProfiler.upload_queue_interval_secs)
+            end
+          end
+          @process_queue_thread.priority = -1 # low priority
+        end
+
         private
 
+        def init_queue
+          @queue = SizedQueue.new(AppProfiler.max_upload_queue_length)
```

Comment:
In process_queue we call `init_queue` unconditionally, so could we possibly race here? Ie, could something add something to the queue, then we call init queue and clobber that object? Perhaps this should be an `||=`

Reply:
related to https://github.com/Shopify/app_profiler/pull/80#discussion_r1223435346 -- we want to reset it every time we process the queue.